File: leetcode.com/swift/DS_ALGO_PRAC.playground/Sources/swift/150_Evaluate_Reverse_Polish_Notation.swift
import Foundation

class Solution {
    func evalRPN(_ tokens: [String]) -> Int {
        var opes = ["+", "-", "*", "/"]
        var stack = [String]()
        for token in tokens {
            if opes.contains(token)  {
                var rightOperand: Double! = Double(stack.popLast()!)
                var leftOperand: Double! = Double(stack.popLast()!)
                var result = 0.0
                if token == "+" {
                    result = leftOperand + rightOperand
                } else if token == "-" {
                    result = leftOperand - rightOperand
                } else if token == "*" {
                    result = leftOperand * rightOperand
                } else {
                    if (rightOperand*leftOperand < 0) && (Int(leftOperand)%Int(rightOperand) != 0) {
                        result = leftOperand / rightOperand + 1
                    } else {
                        result = leftOperand / rightOperand
                    }
                }
                var res = Int(result.round(.towardZero)
                stack.append(String(res))
            } else {
                stack.append(token)
            }
        }
        print("stack: \(stack)")
        return Int(stack.popLast()!)!
    }
}

File: leetcode.com/python/150_Evaluate_Reverse_Polish_Notation.py
# https://tinyurl.com/r35ffgt
from collections import deque
class Solution(object):
    def evalRPN(self, tokens):
        """
        :type tokens: List[str]
        :rtype: int
        """
        tokensDeque = deque(tokens)
        stack = []
        opes = ["+", "-", "*", "/"]
        while tokensDeque:
            currentToken = tokensDeque.popleft()
            if currentToken in opes:
                firstNum = int(stack.pop())
                secondNum = int(stack.pop())
                result = 0
                if currentToken == "+":
                    result = secondNum + firstNum
                elif currentToken == "-":
                    result = secondNum - firstNum
                elif currentToken == "*":
                    result = secondNum * firstNum
                else:
                    if firstNum*secondNum < 0 and secondNum%firstNum != 0:
                        result = secondNum // firstNum + 1
                    else:
                        result = secondNum // firstNum
                stack.append(str(result))
            else:
                stack.append(currentToken)
        return stack[-1]
